content/Patterns for maintainable Rails applications. Part I - Ignacio Pérez.md

```diff
@@ -0,0 +1,111 @@
+
+#### Query Objects
+
+These are Plain Old Ruby Objects (PORO) that include a model’s query logic. The goal of this pattern is to reduce the responsibility of a model by removing complex queries from it. There are two different versions for this pattern: one in which we create one object per model, and another where we create one per scope we want to extract. The second version will likely be the best choice if we have multiple complex queries.
+
+Imagine we had a e-commerce application and we regularly send emails to users that haven’t visited us recently and users that recently bought products. We may have then a User model with some scopes, looking like the following:
+
+https://gist.github.com/PerezIgnacio/a767c7a024c44464698fa57c623f2a76
+
+The following is an example of this pattern applied to User model:
+
+https://gist.github.com/PerezIgnacio/cee4e706abf50ed707b05adca61acbca
+
+Then we can use it from somewhere else in our code, for example, in a simple task:
+
+https://gist.github.com/PerezIgnacio/01a45bb385afb84b1a8e792360a30696
+
+We generate a public method for each scope. The query object is initialized with an Active Record relation as an optional parameter, and if there isn’t any given as argument, it initializes with all records of the relation. This also allows composition with other query objects and scopes, similar as if we were using scopes.
+
+As you may have noticed, the subscribed scope behavior is repeated inside the query object. Fortunately, we can encapsulate queries or part of complex ones into auto-descriptive methods. For this, we dynamically extend the relation with “private” scopes and then use them in our query object, using `ActiveRecord::QueryMethods.extending` method in our initializer.
+
+https://gist.github.com/PerezIgnacio/b427f353bd8b8bbeb06be7c9e2b222b2
+
+### Nested attributes
+
+With nested attributes we can save the attributes of a record through an associated record. This makes them useful as we can easily create multiple nested records and automatically handle any error.
+
+One issue with nested associations is that we may need to add additional complexity to our controllers. The biggest problem with them, however, is that it restricts the parameters our controllers must receive, and so it unnecessarily couples the frontend to the database. It's weird for the frontend to know how we designed our models, so we should have an interface for this in the backend. We can achieve this with the Form Objects pattern.
```

Comment:
not sure if it's worth mentioning, but having the frontend tied to the backend is specially weird when the frontend is a completely separated app like React/Ember/Angular .